protocol-units/dispute/src/RStarM.sol

```diff
@@ -0,0 +1,259 @@
+
+    uint256 public constant MIN_STAKE = 1 ether;
+    uint256 public delta = 1 * SECONDS_IN_DAY; // Time window for filing a dispute (e.g., 1 day)
+    uint256 public p = 1 * SECONDS_IN_MINUTE; 
+    uint256 public m; // Minimum number of validators requried
+
+    /// @notice Control ID hash for the identity_p254 predicate decomposed by `splitDigest`.
+    /// @dev This value controls what set of recursion programs, and therefore what version of the
+    /// zkVM circuit, will be accepted by this contract. Each instance of this verifier contract
+    /// will accept a single release of the RISC Zero circuits.
+    ///
+    /// New releases of RISC Zero's zkVM require updating these values. These values can be
+    /// obtained by running `cargo run --bin bonsai-ethereum-contracts -F control-id`
+    uint256 public immutable CONTROL_ID_0;
+    uint256 public immutable CONTROL_ID_1;
+    uint256 public immutable BN254_CONTROL_ID;
+
+    mapping(address => Validator) public validators;
+    mapping(bytes32 => Dispute) public disputes;
+    mapping(bytes32 => Proof) public verifiedProofs; // Maps blockHash to Proof
+    mapping(bytes32 => OptimisticCommitment) public optimisticCommitments; // Maps blockHash to OptimisticCommitment
+
+    IRiscZeroVerifier public verifier;
+
+    event ValidatorRegistered(address indexed validator, uint256 stake);
+    event ValidatorDeregistered(address indexed validator);
+    event DisputeSubmitted(bytes32 indexed disputeHash, bytes32 blockHash, address indexed submitter);
+    event DisputeResolved(bytes32 indexed disputeHash, DisputeState state);
+    event ProofSubmitted(bytes32 indexed blockHash, bool isValid);
+    event ProofVerified(bytes32 indexed blockHash, bool isValid);
+    event BlockAccepted(bytes32 indexed blockHash);
+    event OptimisticCommitmentSubmitted(bytes32 indexed blockHash, bytes stateCommitment, uint256 validatorCount);
+
+    constructor(
+        uint256 _delta, 
+        uint256 _p, 
+        uint256 _m, 
+        uint256 control_id_0, 
+        uint256 control_id_1,
+        uint256 bn254_control_id
+    ) {
+        delta = _delta;
+        p = _p;
+        m = _m;
+        CONTROL_ID_0 = control_id_0;
+        CONTROL_ID_1 = control_id_1;
+        BN254_CONTROL_ID = bn254_control_id;
+    }
+
+    /// @notice splits a digest into two 128-bit words to use as public signal inputs.
+    /// @dev RISC Zero's Circom verifier circuit takes each of two hash digests in two 128-bit
+    /// chunks. These values can be derived from the digest by splitting the digest in half and
+    /// then reversing the bytes of each.
+    function splitDigest(bytes32 digest) internal pure returns (uint256, uint256) {
+        uint256 reversed = reverseByteOrderUint256(uint256(digest));
+        return (uint256(uint128(uint256(reversed))), uint256(reversed >> 128));
+    }
+
+    function registerValidator() external payable {
```

Comment:
It makes even more sense if we going with the staked rollup approach to call these `stake` and `unstake`.